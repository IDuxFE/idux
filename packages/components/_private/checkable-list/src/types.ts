/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
