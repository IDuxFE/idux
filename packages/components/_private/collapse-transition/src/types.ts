/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
