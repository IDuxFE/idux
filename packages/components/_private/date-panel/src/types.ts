/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
