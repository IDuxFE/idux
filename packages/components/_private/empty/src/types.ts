/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
