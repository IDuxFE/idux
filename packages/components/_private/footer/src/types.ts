/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
