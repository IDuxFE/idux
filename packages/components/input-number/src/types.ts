/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { controlPropDef } from '@idux/cdk/forms'
import { IxPropTypes } from '@idux/cdk/utils'

export type InputNumberButtonPosition = 'inner' | 'outer'

export const inputNumberProps = {
  value: IxPropTypes.oneOfType<number | null>([IxPropTypes.number]),
  control: controlPropDef,
  disabled: IxPropTypes.bool.def(false),
  keyboard: IxPropTypes.bool,
  max: IxPropTypes.number.def(Infinity),
  min: IxPropTypes.number.def(-Infinity),
  placeholder: IxPropTypes.string,
  precision: IxPropTypes.number,
  readonly: IxPropTypes.bool.def(false),
  size: IxPropTypes.oneOf<FormSize>(['sm', 'md', 'lg']),
  step: IxPropTypes.number.def(1),

  // events
  'onUpdate:value': IxPropTypes.emit<(value: number | null) => void>(),
  onChange: IxPropTypes.emit<(value: number | null, oldValue: number | null | undefined) => void>(),
  onFocus: IxPropTypes.emit<(evt: FocusEvent) => void>(),
  onBlur: IxPropTypes.emit<(evt: FocusEvent) => void>(),
}

export type InputNumberProps = IxInnerPropTypes<typeof inputNumberProps>
export type InputNumberPublicProps = IxPublicPropTypes<typeof inputNumberProps>
export interface InputNumberBindings {
  focus: (options?: FocusOptions) => void
  blur: () => void
}
export type InputNumberComponent = DefineComponent<
  Omit<HTMLAttributes, keyof InputNumberPublicProps> & InputNumberPublicProps,
  InputNumberBindings
>
export type InputNumberInstance = InstanceType<DefineComponent<InputNumberProps>>
