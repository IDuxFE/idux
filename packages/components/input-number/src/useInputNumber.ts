/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { type ComputedRef, type Ref, computed, nextTick, ref, toRaw, watch } from 'vue'

import { type ValidateStatus, useAccessorAndControl } from '@idux/cdk/forms'
import { Logger, callEmit } from '@idux/cdk/utils'
import { type InputNumberConfig } from '@idux/components/config'
import { type FormSize, useFormItemRegister, useFormSize, useFormStatus } from '@idux/components/form'

import { type InputNumberProps } from './types'

export interface InputNumberBindings {
  mergedSize: ComputedRef<FormSize | undefined>
  mergedStatus: ComputedRef<ValidateStatus | undefined>
  displayValue: Ref<string>
  isIllegal: Ref<boolean>
  isDisabled: ComputedRef<boolean>
  isFocused: Ref<boolean>
  nowValue: ComputedRef<number | undefined>
  isDisabledDec: ComputedRef<boolean>
  isDisabledInc: ComputedRef<boolean>

  handleKeyDown: (evt: KeyboardEvent) => void
  handleDec: () => void
  handleInc: () => void
  handleInput: (evt: Event) => void
  handleFocus: (evt: FocusEvent) => void
  handleBlur: (evt: FocusEvent) => void
}

export function useInputNumber(props: InputNumberProps, config: InputNumberConfig): InputNumberBindings {
  const { accessor, control } = useAccessorAndControl<number | null>()
  useFormItemRegister(control)
  const mergedSize = useFormSize(props, config)
  const mergedStatus = useFormStatus(props, control)

  const displayValue = ref('')
  const isIllegal = ref(true)
  const nowValue = computed(() => accessor.value ?? undefined)
  const isKeyboard = computed(() => props.keyboard ?? config.keyboard)
  const isDisabled = computed(() => accessor.disabled)
  const isDisabledDec = computed(() => props.readonly || (!!nowValue.value && nowValue.value <= props.min))
  const isDisabledInc = computed(() => props.readonly || (!!nowValue.value && nowValue.value >= props.max))

  const precision = computed(() => {
    const stepPrecision = getPrecision(props.step)
    if (props.precision !== undefined) {
      if (__DEV__ && stepPrecision > props.precision) {
        Logger.warn(
          'components/input-number',
          `The precision(${props.precision}) should not be less than the decimal point of the step(${props.step}).`,
        )
      }
      return props.precision
    }
    return Math.max(getPrecision(accessor.value), stepPrecision)
  })

  const disabledDec = computed(() => getIncValueFormAccessor(-props.step) < props.min)
  const disabledInc = computed(() => getIncValueFormAccessor(props.step) > props.max)

  function getIncValueFormAccessor(step: number) {
    const { value } = accessor
    let newVal = step
    if (typeof value === 'number' && !Number.isNaN(value)) {
      // Use the toFixed func to ensure numerical accuracy.
      newVal = parseFloat((value + step).toFixed(precision.value))
    }

    return Math.max(props.min, Math.min(props.max, newVal))
  }

  function updateDisplayValueFromAccessor() {
    const { value } = accessor
    if (value === null || value === undefined) {
      displayValue.value = ''
    } else if (Number.isNaN(value)) {
      displayValue.value = ''
      if (__DEV__) {
        Logger.warn('components/input-number', `model value(${value}) is not a number.`)
      }
    } else {
      if (displayValue.value === '' || value !== Number(displayValue.value)) {
        displayValue.value = value.toFixed(precision.value)
      }
    }
  }

  function updateModelValueFromDisplayValue() {
    const { value: strVal } = displayValue
    const numberVal = parseFloat(Number(strVal).toFixed(precision.value))

    // When the model value is `NaN`, value of input element is empty string, update the model value to `null`.
    if (strVal === '') {
      updateModelValue(null)
    } else if (Number.isNaN(numberVal)) {
      updateDisplayValueFromAccessor()
    } else {
      const newVal = Math.max(props.min, Math.min(props.max, numberVal))
      displayValue.value = newVal.toFixed(precision.value)
      updateModelValue(newVal)
    }
  }

  function updateModelValue(newVal: number | null) {
    const oldVal = toRaw(accessor.value)
    if (newVal !== oldVal) {
      accessor.setValue(newVal)
      callEmit(props.onChange, newVal, oldVal)

      nextTick(() => {
        if (newVal !== accessor.value) {
          updateDisplayValueFromAccessor()
        }
      })
    }
  }

  function handleInput(evt: Event) {
    const { value: inputVal } = evt.target as HTMLInputElement
    const strVal = inputVal.trim().replace(/ã€‚/g, '.')

    displayValue.value = strVal

    if (strVal === '') {
      updateModelValue(null)
      return
    }

    const numberVal = Number(strVal)
    if (!Number.isNaN(numberVal)) {
      if (numberVal >= props.min && numberVal <= props.max) {
        updateModelValue(numberVal)
      }
    }
  }

  function handleDec() {
    if (props.readonly || isDisabled.value || disabledDec.value) {
      return
    }

    const newVal = getIncValueFormAccessor(-props.step)
    updateModelValue(newVal)
  }

  function handleInc() {
    if (props.readonly || isDisabled.value || disabledInc.value) {
      return
    }

    const newVal = getIncValueFormAccessor(props.step)
    updateModelValue(newVal)
  }

  function handleKeyDown(evt: KeyboardEvent): void {
    if (isKeyboard.value) {
      if (evt.code === 'Enter' || evt.code === 'NumpadEnter') {
        updateModelValueFromDisplayValue()
      } else if (evt.code === 'ArrowUp') {
        evt.preventDefault()
        handleInc()
      } else if (evt.code === 'ArrowDown') {
        evt.preventDefault()
        handleDec()
      }
    }
  }

  const isFocused = ref(false)
  function handleFocus(evt: FocusEvent) {
    isFocused.value = true
    callEmit(props.onFocus, evt)
  }

  function handleBlur(evt: FocusEvent) {
    isFocused.value = false
    updateModelValueFromDisplayValue()
    accessor.markAsBlurred()
    callEmit(props.onBlur, evt)
  }

  watch(
    displayValue,
    value => {
      if (value !== '') {
        const numberVal = Number(value)
        isIllegal.value = Number.isNaN(numberVal) || numberVal < props.min || numberVal > props.max
      } else {
        isIllegal.value = false
      }
    },
    { immediate: true },
  )

  watch(
    () => accessor.value,
    () => updateDisplayValueFromAccessor(),
    { immediate: true },
  )

  return {
    mergedSize,
    mergedStatus,
    displayValue,
    isIllegal,
    isDisabled,
    isDisabledDec,
    isDisabledInc,
    isFocused,
    nowValue,
    handleKeyDown,
    handleDec,
    handleInc,
    handleInput,
    handleFocus,
    handleBlur,
  }
}

export function getPrecision(value: number | undefined | null): number {
  if (value === undefined || value === null) {
    return 0
  }

  const decimal = String(value).split('.')[1]
  return decimal ? decimal.length : 0
}
