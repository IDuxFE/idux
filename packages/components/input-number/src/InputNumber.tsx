/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ɵInputInstance } from '@idux/components/_private/input'

import { computed, defineComponent, normalizeClass, onMounted, ref } from 'vue'

import { ɵInput } from '@idux/components/_private/input'
import { useGlobalConfig } from '@idux/components/config'
import { useFormFocusMonitor, useFormSize } from '@idux/components/form'
import { IxIcon } from '@idux/components/icon'

import { inputNumberProps } from './types'
import { useInputNumber } from './useInputNumber'

export default defineComponent({
  name: 'IxInputNumber',
  props: inputNumberProps,
  setup(props, { expose, slots }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('inputNumber')
    const {
      displayValue,
      nowValue,
      isIllegal,
      isDisabled,
      isDisabledDec,
      isDisabledInc,
      isFocused,
      handleInput,
      handleFocus,
      handleBlur,
      handleKeyDown,
      handleDec,
      handleInc,
    } = useInputNumber(props, config)
    const { elementRef, focus, blur } = useFormFocusMonitor<HTMLInputElement>({ handleBlur, handleFocus })
    expose({ focus, blur })

    const mergedPrefixCls = computed(() => `${common.prefixCls}-input-number`)
    const mergedSize = useFormSize(props, config)

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-illegal`]: isIllegal.value,
      })
    })

    const inputRef = ref<ɵInputInstance>()
    onMounted(() => {
      elementRef.value = inputRef.value!.getInputElement()
    })

    const handleFocusAtClick = () => {
      if (!isDisabled.value) {
        focus()
      }
    }

    const handleIncrease = () => {
      handleInc()
      handleFocusAtClick()
    }

    const handleDecrease = () => {
      handleDec()
      handleFocusAtClick()
    }

    const vSlots = computed(() => {
      return {
        ...slots,
        prefix: () => (
          <span
            class={['ix-input-number-increase', isDisabledInc.value ? 'ix-input-number-increase--disabled' : '']}
            role="button"
            onClick={handleIncrease}
          >
            <IxIcon name="up" />
          </span>
        ),
        suffix: () => (
          <span
            class={['ix-input-number-decrease', isDisabledDec.value ? 'ix-input-number-decrease--disabled' : '']}
            role="button"
            onClick={handleDecrease}
          >
            <IxIcon name="down" />
          </span>
        ),
      }
    })

    return () => {
      return (
        <ɵInput
          class={classes.value}
          ref={inputRef}
          type="text"
          autocomplete="off"
          aria-valuemin={props.min}
          aria-valuemax={props.max}
          aria-valuenow={nowValue.value}
          disabled={isDisabled.value}
          focused={isFocused.value}
          readonly={props.readonly}
          placeholder={props.placeholder}
          size={mergedSize.value}
          value={displayValue.value}
          onInput={handleInput}
          onKeydown={handleKeyDown}
          v-slots={vSlots.value}
        />
      )
    }
  },
})
