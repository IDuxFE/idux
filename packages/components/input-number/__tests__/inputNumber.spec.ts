import { MountingOptions, mount } from '@vue/test-utils'
import { nextTick, ref } from 'vue'

import { renderWork } from '@tests'

import InputNumber from '../src/InputNumber'
import { InputNumberProps } from '../src/types'

describe('InputNumber', () => {
  const InputNumberMount = (options?: MountingOptions<Partial<InputNumberProps>>) => mount(InputNumber, { ...options })

  renderWork(InputNumber)

  test('v-model work', async () => {
    const valueRef = ref(1)
    const wrapper = mount({
      components: { InputNumber },
      template: `<InputNumber v-model:value="valueRef" />`,
      setup() {
        return { valueRef }
      },
    })
    const input = wrapper.find('input')

    expect(input.element.value).toBe(valueRef.value.toString())

    await input.setValue(10)

    expect(valueRef.value).toBe(10)

    valueRef.value = 5
    await nextTick()

    expect(input.element.value).toBe(valueRef.value.toString())
  })

  test('controlled value work', async () => {
    const wrapper = InputNumberMount({ props: { value: 1 } })
    const input = wrapper.find('input')

    expect(input.element.value).toBe('1')

    await input.setValue(10)
    await nextTick()

    expect(input.element.value).toBe('1')
  })

  test('disabled work', async () => {
    const onUpdateValue = jest.fn()
    const wrapper = InputNumberMount({ props: { disabled: true, 'onUpdate:value': onUpdateValue } })

    await wrapper.find('.ix-input-number-increase').trigger('click')

    expect(onUpdateValue).not.toBeCalled()

    await wrapper.setProps({ disabled: false })
    await wrapper.find('.ix-input-number-increase').trigger('click')

    expect(onUpdateValue).toBeCalled()
  })

  test('readonly work', async () => {
    const onUpdateValue = jest.fn()
    const wrapper = InputNumberMount({ props: { readonly: true, 'onUpdate:value': onUpdateValue } })

    await wrapper.find('.ix-input-number-increase').trigger('click')

    expect(onUpdateValue).not.toBeCalled()

    await wrapper.setProps({ readonly: false })
    await wrapper.find('.ix-input-number-increase').trigger('click')

    expect(onUpdateValue).toBeCalled()
  })

  test('step work', async () => {
    const valueRef = ref(1)
    const wrapper = mount({
      components: { InputNumber },
      template: `<InputNumber v-model:value="valueRef" />`,
      setup() {
        return { valueRef }
      },
    })

    await wrapper.find('.ix-input-number-decrease').trigger('click')

    expect(valueRef.value).toBe(0)

    await wrapper.find('.ix-input-number-increase').trigger('click')

    expect(valueRef.value).toBe(1)
  })

  test('precision work', async () => {
    const wrapper = InputNumberMount({ props: { value: 1, precision: 2 } })
    const input = wrapper.find('input')

    expect(input.element.value).toBe('1.00')
  })

  test('keyboard work', async () => {
    const valueRef = ref(1)
    const wrapper = mount({
      components: { InputNumber },
      template: `<InputNumber v-model:value="valueRef" :keyboard="false" />`,
      setup() {
        return { valueRef }
      },
    })
    const input = wrapper.find('input')

    await input.trigger('focus')
    await input.trigger('keydown', { code: 'ArrowUp' })

    expect(input.element.value).toBe('1')
    expect(valueRef.value).toBe(1)

    await input.trigger('keydown', { code: 'ArrowDown' })

    expect(input.element.value).toBe('1')
    expect(valueRef.value).toBe(1)

    await wrapper.setProps({ keyboard: true })

    await input.trigger('keydown', { code: 'ArrowUp' })

    expect(input.element.value).toBe('2')
    expect(valueRef.value).toBe(2)

    await input.trigger('keydown', { code: 'ArrowDown' })

    expect(input.element.value).toBe('1')
    expect(valueRef.value).toBe(1)
  })
})
