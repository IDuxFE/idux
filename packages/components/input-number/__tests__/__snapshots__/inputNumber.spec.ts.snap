// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputNumber render work 1`] = `"<span class=\\"ix-input ix-input-md ix-input-with-addon-after ix-input-with-addon-before ix-input-number\\"><span class=\\"ix-input-addon\\"><span class=\\"ix-input-number-decrease\\" role=\\"button\\"><i class=\\"ix-icon ix-icon-minus\\" role=\\"img\\" aria-label=\\"minus\\"></i></span></span><input class=\\"ix-input-inner\\" type=\\"text\\" autocomplete=\\"off\\" aria-valuemin=\\"-Infinity\\" aria-valuemax=\\"Infinity\\"><span class=\\"ix-input-addon\\"><span class=\\"ix-input-number-increase\\" role=\\"button\\"><i class=\\"ix-icon ix-icon-plus\\" role=\\"img\\" aria-label=\\"plus\\"></i></span></span></span>"`;
